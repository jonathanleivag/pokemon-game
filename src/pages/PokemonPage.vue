<template>
  <h1 v-if="!pokemon">Cargando....</h1>
  <div v-else>
    <h1>¿Quién es este Pokémon?</h1>
    <PokemonPictureVue :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptionsVue :pokemons="pokemons" @selector="checkAnswer" :toDisable="toDisable" />
    <template v-if="showAswer">
      <h2 class="fade-in"> {{message}} </h2>
      <button @click="next">Siguiente</button>
    </template>
  </div>
</template>

<script>
import PokemonOptionsVue from '../components/PokemonOptions.vue'
import PokemonPictureVue from '../components/PokemonPicture.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components:{
        PokemonPictureVue,
        PokemonOptionsVue
    },
    data(){
      return {
        pokemons: [],
        pokemon: null,
        showPokemon: false,
        showAswer: false,
        message: null,
        toDisable: false
      }
    },
    methods: {
      async mixPokemonArray(){
        this.pokemons = await getPokemonOptions()
        const pokemonRandom = window.Math.floor(window.Math.random() * 4)
        this.pokemon = this.pokemons[pokemonRandom]
      },
      checkAnswer(pokemonId){
        this.showPokemon = true
        this.showAswer= true
        this.toDisable= true
        if(pokemonId === this.pokemon.id){
          this.message = `Correcto, es ${this.pokemon.name}`
        }else {
          this.message = `Oops, era ${this.pokemon.name}`
        }
      },
       next(){
        this.showPokemon = false
        this.showAswer= false
        this.pokemon = null
        this.toDisable= false
        this.pokemons = []
         this.mixPokemonArray()
       }
    },
    mounted(){
      this.mixPokemonArray()
    }
}
</script>

<style>

</style>