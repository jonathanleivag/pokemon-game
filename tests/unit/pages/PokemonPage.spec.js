import PokemonPage from '@/pages/PokemonPage'
import { shallowMount } from '@vue/test-utils'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage', () => {
  it('Inicio de componente', async () => {
    const wrapper = shallowMount(PokemonPage)
    const h1 = wrapper.find('h1')
    expect(h1.text()).toBe('Cargando....')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('debe llamar a "mixPokemonArray" al montar', async () => {
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      'mixPokemonArray'
    )
    shallowMount(PokemonPage)

    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })

  it('debe hacer mach con el snapshot', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAswer: false,
          message: null,
          toDisable: false
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('debe mestrar "PokemonPictureVue" y "PokemonOptionsVue" con sus atributos', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAswer: false,
          message: null,
          toDisable: false
        }
      }
    })

    const picture = wrapper.find('pokemon-picture-vue-stub')
    expect(picture.exists()).toBe(true)
    expect(picture.attributes('pokemonid')).toBe(pokemons[0].id.toString())
    expect(picture.attributes('showpokemon')).toBe('false')

    const options = wrapper.find('pokemon-options-vue-stub')
    expect(options.exists()).toBe(true)
    expect(options.attributes('todisable')).toBe('false')
    expect(options.attributes('pokemons')).toBeTruthy()
    expect(options.attributes('pokemons')).toStrictEqual(expect.any(String))
  })

  it('checkAnswer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAswer: false,
          message: null,
          toDisable: false
        }
      }
    })

    await wrapper.vm.checkAnswer(1)

    expect(wrapper.vm.showPokemon).toBe(true)
    expect(wrapper.vm.showAswer).toBe(true)
    expect(wrapper.vm.toDisable).toBe(true)
    expect(1).toBe(wrapper.vm.pokemon.id)
    expect(wrapper.vm.message).toBe(`Correcto, es ${wrapper.vm.pokemon.name}`)

    await wrapper.vm.checkAnswer(2)
    expect(2).not.toBe(wrapper.vm.pokemon.id)
    expect(wrapper.vm.message).toBe(`Oops, era ${wrapper.vm.pokemon.name}`)

    const wrapper0 = shallowMount(PokemonPage, {
      data () {
        return {
          pokemons,
          pokemon: pokemons[1],
          showPokemon: false,
          showAswer: false,
          message: null,
          toDisable: false
        }
      }
    })

    await wrapper0.vm.checkAnswer(2)
    expect(2).toBe(wrapper0.vm.pokemon.id)
  })

  it('boton siguiente', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: true,
          showAswer: true,
          message: '',
          toDisable: true
        }
      }
    })

    wrapper.vm.checkAnswer(1)

    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      'mixPokemonArray'
    )

    await wrapper.find('button').trigger('click')

    expect(wrapper.vm.showPokemon).toBe(false)
    expect(wrapper.vm.showAswer).toBe(false)
    expect(wrapper.vm.toDisable).toBe(false)
    expect(wrapper.vm.pokemons).toStrictEqual([])
    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })
})
