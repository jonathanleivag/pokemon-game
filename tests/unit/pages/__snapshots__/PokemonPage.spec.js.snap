// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Inicio de componente 1`] = `<h1>Cargando....</h1>`;

exports[`PokemonPage debe hacer mach con el snapshot 1`] = `
<div>
  <h1>¿Quién es este Pokémon?</h1>
  <pokemon-picture-vue-stub pokemonid="1" showpokemon="false"></pokemon-picture-vue-stub>
  <pokemon-options-vue-stub pokemons="[object Object],[object Object],[object Object],[object Object]" todisable="false"></pokemon-options-vue-stub>
  <!--v-if-->
</div>
`;
